### ISSUE TITLE
Is it possible to use Vtk9.0.0

### ISSUE THREAD

**Issue by crocketyang** [qCring/QuickVtk#56]
Is it possible to use Vtk9.0.0, since I know Vtk9.0.0 fixed many issues.

**Comment by maidamai0** [comment]
Did't compile with vtk9.0.0.
I just compiled it with vtk from [master](https://gitlab.kitware.com/vtk/vtk), after fixing some small issues there is one left I don't know how to do at quickVtkFboOffscreenWindow.cpp:
```cpp
auto FboOffscreenWindow::SetFramebufferObject(QOpenGLFramebufferObject *fbo) -> void {
    this->BackLeftBuffer = this->FrontLeftBuffer = this->BackBuffer = this->FrontBuffer = static_cast<unsigned int>(GL_COLOR_ATTACHMENT0);

    auto size                    = fbo->size();

    this->Size[0]                   = size.width();
    this->Size[1]                   = size.height();
    this->NumberOfFrameBuffers      = 1;
    this->FrameBufferObject         = static_cast<unsigned int>(fbo->handle());
    this->DepthRenderBufferObject   = 0;
    this->TextureObjects[0]         = static_cast<unsigned int>(fbo->texture());
    this->OffScreenRendering        = 1;
    this->OffScreenUseFrameBuffer   = 1;
    this->Modified();
}
```
So, from [this commit](https://gitlab.kitware.com/vtk/vtk/-/commit/588247020e455b9c3f7bf37a64525c31c5071f52) some real changes must be made to compile with the latest vtk.

**Comment by maidamai0** [comment]
According to this issue on [gitlab vtk](https://gitlab.kitware.com/vtk/vtk/-/issues/16061#note_788163) it seems still possible to to this.But I don't know how.

**Comment by danielbuhrig** [comment]
Any solution yet? I'm facing the same problem.

**Comment by qCring** [comment]
Hey @danielbuhrig, I'm working on refactoring the project and set things up for VTK 9.0 but that will take some time. I'll keep this issue updated as development goes on

**Comment by danielbuhrig** [comment]
Cool, thank you!

**Comment by jhuels** [comment]
@qCring Did you make any progress here? I'm also at looking integrating using the `QQuickFrameBuffer` interface.  From your code, I believe the `SetFrameBufferObject` function needs to become something like the following:

```cpp
    auto FboOffscreenWindow::SetFramebufferObject(QOpenGLFramebufferObject* fbo) -> void
    {
        vtkTextureObject* tex = vtkTextureObject::New();
        tex->AssignToExistingTexture(fbo->handle(), fbo->texture());
        this->GetDisplayFramebuffer()->AddColorAttachment(0, tex);
        this->Modified();
    }
```

After making this change, I'm having some problems with initializing the OpenGL state.  

In `OpenGLInitState` the application is getting an Access Violation in `vtkOpenGLState::Initialize()` on the following function `::glBlendFuncSeparate(cs.BlendFunc[0], cs.BlendFunc[1], cs.BlendFunc[2], cs.BlendFunc[3]);`.

Let me know if you've seen this, maybe we can help each other out.

**EDIT**: If I call `glBlendFuncSeparate` directly in `OpenGLInitState`, it seems to take fine.   But the code snippet above probably isn't right.  I'm getting lots of gl errors now:

```
vtkTextureObject.cxx:1751   ERR| vtkTextureObject (000001E44367A6E0): failed at glTexImage2D 1 OpenGL errors detected
  0 : (1280) Invalid enum
```

I'm just going to roll back to 8.2 for now.

**Comment by JensMunkHansen** [comment]
Hi Jhuels

The initialization you get is most likely due to Qt initializing the
context using the message loop.

I had similar issues with this before and polled the application - process
events until the window is valid. If not all glXXX calls will segfault.

We are still stuck with 8.2 for other reasons, but I will soon move to 9.0

/Jens

**Comment by jhuels** [comment]
Thanks for the hint Jens!

I found out it was due to GLEW not being initialized before setting up the context state.  This can be solved by calling `vtkOpenGLRenderWindow::OpenGLInitContext()` before initializing the `vtkOpenGLState`.

**Comment by behtgod** [comment]
@jhuels Did you means that you have success to migrate to VTK 9.0?

**Comment by jhuels** [comment]
No, I just used 8.2 cause it worked.  I had the same issue on both versions, so presumably you could get 9.0 working as well.
